c++/src/OrcHdfsFile.cc

```diff
@@ -0,0 +1,170 @@
+      if(config){
+        //Loading options from the config
+        options = config->GetOptions();
+      }
+      hdfs::IoService * io_service = hdfs::IoService::New();
+      //Wrapping fs into a shared pointer to guarantee deletion
+      std::shared_ptr<hdfs::FileSystem> fs(hdfs::FileSystem::New(io_service, "", options));
+      if (!fs) {
+        throw ParseError("Can't create FileSystem object. ");
+      }
+      hdfs::Status status;
+      //Check if the user supplied the host
+      if(!uri.value().get_host().empty()){
+        //If port is supplied we use it, otherwise we use the empty string so that it will be looked up in configs.
+        std::string port = (uri.value().get_port()) ? std::to_string(uri.value().get_port().value()) : "";
+        status = fs->Connect(uri.value().get_host(), port);
+        if (!status.ok()) {
+          throw ParseError("Can't connect to " + uri.value().get_host() + ":" + port + ". " + status.ToString());
+        }
+      } else {
+        status = fs->ConnectToDefaultFs();
+        if (!status.ok()) {
+          if(!options.defaultFS.get_host().empty()){
+            throw ParseError("Error connecting to " + options.defaultFS.str() + ". " + status.ToString());
+          } else {
+            throw ParseError("Error connecting to the cluster: defaultFS is empty. " + status.ToString());
+          }
+        }
+      }
+
+      if (!fs) {
+        throw ParseError("Can't connect the file system. ");
+      }
+
+      hdfs::FileHandle *file_raw = nullptr;
+      status = fs->Open(uri->get_path(), &file_raw);
+      if (!status.ok()) {
+        throw ParseError("Can't open " + uri->get_path() + ". " + status.ToString());
+      }
+      //wrapping file_raw into a unique pointer to guarantee deletion
+      std::unique_ptr<hdfs::FileHandle> file_handle(file_raw);
+      file = std::move(file_handle);
+      //transferring the ownership of FileSystem to HdfsFileInputStream to avoid premature deallocation
+      file_system = fs;
+
+      hdfs::StatInfo stat_info;
+      status = fs->GetFileInfo(uri->get_path(), stat_info);
+      if (!status.ok()) {
+        throw ParseError("Can't stat " + uri->get_path() + ". " + status.ToString());
+      }
+      totalLength = stat_info.length;
+    }
+
+    ~HdfsFileInputStream();
+
+    uint64_t getLength() const override {
+      return totalLength;
+    }
+
+    uint64_t getNaturalReadSize() const override {
+      return BUF_SIZE;
+    }
+
+    void read(void* buf,
+              uint64_t length,
+              uint64_t offset) override {
+
+      ssize_t total_bytes_read = 0;
+      size_t last_bytes_read = 0;
+
+      if (!buf) {
+        throw ParseError("Buffer is null");
+      }
+
+      hdfs::Status status;
+      char input_buffer[BUF_SIZE];
+      do{
+        //Reading file chunks
+        status = file->PositionRead(input_buffer, sizeof(input_buffer), offset, &last_bytes_read);
```

Comment:
I used intermediate buffer because it was used for PositionRead in a similar case. So I just tried without this buffer and it worked.